### PR TITLE
LibJS: Some copy elisions and MVL interface niceties

### PR DESCRIPTION
**Pull request by alimpfard**
This commit also eliminates quite a few unnecessary copies of Vectors.

Why redo Vector features, or forward calls when we can just inherit from Vector

cc @linusg - in case I messed some required copy up.

### CODE REVIEW

File: Libraries/LibJS/Runtime/IndexedProperties.cpp [hunk 1]
```diff
@@ -374,7 +374,7 @@ Vector<ValueAndAttributes> IndexedProperties::values_unordered() const
 
 void IndexedProperties::switch_to_generic_storage()
 {
-    auto storage = static_cast<const SimpleIndexedPropertyStorage&>(*m_storage);
+    auto& storage = static_cast<SimpleIndexedPropertyStorage&>(*m_storage);
```
Comment (alimpfard) [inline Libraries/LibJS/Runtime/IndexedProperties.cpp:377]: "necessary copy": In particular, this one.

File: Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 2]
```diff
@@ -558,7 +558,7 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::reverse)
     if (array->indexed_properties().is_empty())
         return array;
 
-    Vector<Value> array_reverse;
+    MarkedValueList array_reverse{array->heap()};
```
Comment (linusg) [inline Libraries/LibJS/Runtime/ArrayPrototype.cpp:561]:
```suggestion
    MarkedValueList array_reverse { array->heap() };
```

at least I think that's what clang-format should do?

Comment (alimpfard) [reply Libraries/LibJS/Runtime/ArrayPrototype.cpp:561]: it should lol

**Review by linusg — APPROVED** [review]
> This commit also eliminates quite a few unnecessary copies of Vectors.

I'm definitely responsible for a few of those!

LGTM - except for the changed constructors, I don't understand those well enough to judge :)

File: Libraries/LibJS/Runtime/Object.h [hunk 3]
```diff
@@ -169,7 +169,7 @@ class Object : public Cell {
 
     bool m_is_extensible { true };
     Shape* m_shape { nullptr };
-    Vector<Value> m_storage;
+    MarkedValueList m_storage;
```
Comment (awesomekling) [inline Libraries/LibJS/Runtime/Object.h:172]: This change is not needed, as `JS::Object` will mark the values in its storage in `Object::visit_children()`

File: Libraries/LibJS/Runtime/IndexedProperties.h [hunk 4]
```diff
@@ -83,13 +87,13 @@ class SimpleIndexedPropertyStorage final : public IndexedPropertyStorage {
...
     virtual bool is_simple_storage() const override { return true; }
-    Vector<Value> elements() const { return m_packed_elements; }
+    const MarkedValueList& elements() const { return m_packed_elements; }
 
 private:
     friend GenericIndexedPropertyStorage;
 
     size_t m_array_size { 0 };
-    Vector<Value> m_packed_elements;
+    MarkedValueList m_packed_elements;
```
Comment (awesomekling) [inline Libraries/LibJS/Runtime/IndexedProperties.h:96]: This change is not needed, `Object` will mark these in `visit_children()`.

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Libraries/LibJS/Runtime/Object.h (hunk 3 above)
Comment (alimpfard) [reply Libraries/LibJS/Runtime/Object.h:172]: Should I make `Object` _not_ mark these, or just revert this (and the one below)?

Comment (awesomekling) [reply Libraries/LibJS/Runtime/Object.h:172]: Just revert it :)
We have to visit the objects anyway, so let's avoid tracking more MVL's

File: Libraries/LibJS/Runtime/Function.h [hunk 5]
```diff
@@ -66,12 +66,12 @@ class Function : public Object {
...
 protected:
     explicit Function(Object& prototype);
-    Function(Object& prototype, Value bound_this, Vector<Value> bound_arguments);
+    Function(Object& prototype, Value bound_this, MarkedValueList bound_arguments);
 
 private:
     virtual bool is_function() const override { return true; }
     Value m_bound_this;
-    Vector<Value> m_bound_arguments;
+    MarkedValueList m_bound_arguments;
```
Comment (awesomekling) [inline Libraries/LibJS/Runtime/Function.h:74]: This is also already marked by `Function::visit_children()`.

File: Libraries/LibJS/Interpreter.h [hunk 6]
```diff
@@ -59,9 +59,25 @@ struct ScopeFrame {
...
+        , this_value(move(this_value))
+        , arguments(move(arguments))
+        , environment(environment)
+    {
+    }
+
     FlyString function_name;
     Value this_value;
-    Vector<Value> arguments;
+    MarkedValueList arguments;
```
Comment (awesomekling) [inline Libraries/LibJS/Interpreter.h:80]: This is also already marked by `Interpreter::gather_roots()`.

**Review by awesomekling — CHANGES_REQUESTED** [review]
Moar:

File: Libraries/LibJS/Runtime/Function.h (hunk 5 above)
Comment (alimpfard) [reply Libraries/LibJS/Runtime/Function.h:74]: Hmmm...at this point, I question: what _isn't_ marked and is stored in a `Vector<Value>`?

The commit message certainly no longer applies.

Comment (awesomekling) [reply Libraries/LibJS/Runtime/Function.h:74]: Hmm. The stuff returned from `IndexedProperties::values_unordered()` isn't marked.

Comment (awesomekling) [reply Libraries/LibJS/Runtime/Function.h:74]: `SimpleIndexedPropertyStorage::elements()` returns a vector by copy (for no reason I think), which won't be marked

Comment (awesomekling) [reply Libraries/LibJS/Runtime/Function.h:74]: All the `ValueAndAttributes` usage looks really suspicious and unmarked actually

File: Libraries/LibJS/Interpreter.h (hunk 6 above)
Comment (alimpfard) [reply Libraries/LibJS/Interpreter.h:80]: So is everything in a `CallStack`.

File: Libraries/LibJS/Runtime/Function.h (hunk 5 above)
Comment (alimpfard) [reply Libraries/LibJS/Runtime/Function.h:74]: Aside from the odd copies, it looks like they _are_ marked, through `Object::visit_children()`, `values_unordered()`.

Comment (alimpfard) [reply Libraries/LibJS/Runtime/Function.h:74]: It's a bit of a maze to go through, but I don't think your forgotten objects are here...

**Pull request merged**